source/common/http/hash_policy.cc

```diff
@@ -39,14 +42,40 @@ class HeaderHashMethod : public HashMethodImplBase {
                                     const StreamInfo::FilterStateSharedPtr) const override {
     absl::optional<uint64_t> hash;
 
-    // TODO(mattklein123): Potentially hash on all headers.
     const auto header = headers.get(header_name_);
     if (!header.empty()) {
+      bool check_multiple_values =
+          Runtime::runtimeFeatureEnabled("envoy.reloadable_features.hash_multiple_header_values");
+      absl::InlinedVector<std::string, 1> rewritten_header_values;
+      absl::InlinedVector<absl::string_view, 1> header_values;
+
+      size_t header_size = 1;
+      if (check_multiple_values) {
+        header_size = header.size();
+        header_values.reserve(header_size);
+      }
+
+      for (size_t i = 0; i < header_size; i++) {
+        header_values.push_back(header[i]->value().getStringView());
+      }
+
       if (regex_rewrite_ != nullptr) {
-        hash = HashUtil::xxHash64(regex_rewrite_->replaceAll(header[0]->value().getStringView(),
-                                                             regex_rewrite_substitution_));
+        rewritten_header_values.reserve(header_size);
+        for (auto& value : header_values) {
+          rewritten_header_values.push_back(
+              regex_rewrite_->replaceAll(value, regex_rewrite_substitution_));
+          value = rewritten_header_values.back();
+        }
+      }
+
+      if (check_multiple_values) {
```

Comment:
I would just check num_headers_to_hash > 1 here, though I'm not even sure it's worth it. I think the sort call is going to be a nop/almost free if the size is 1. I would probably just collapse all of this logic?

Reply:
One thing may be worth concern about, the absl::Hash and xxHash64 are using two different hash algorithms. Currently using the `check_multiple_values` to switch two code path totally, so it won't happened to mix the using of those two different hash algorithms.  If we check the `num_headers_to_hash`, then we mix the usage of those two hash algo, so would it be a problem for the value distribution?

Reply:
Ah yes, great call out and sorry I mixed this. In general I would not recommend using absl::Hash for this because they make no guarantees about hash stability. I think for this use case guaranteeing stability is very important and we should use xxHash even for multiple values. Can you make that work? I think just by re-hashing with the additional values?

Reply:
This was my suggestion to use absl::Hash. I've tried using xxHash to hash a vector of strings and @soulxu did a nice job of implementing that in an earlier version of this PR. In my attempts to use it earlier I found it slow -- I think it might have been allocating memory. I was thinking LB computations might be performance sensitive.

@mattklein123 what do you mean in this context by hash stability? You mean when we upgrade absl we'll get different values? Is that an operational issue during upgrades?

I was thinking that perf was important but getting same results with new versions was not. I might have had that wrong.

Reply:
I did some test today also. Basically loop 10000000 times to do the hash (not sure it is right way to benchmark)

for absl::Hash
`for (int i = 0; i < count; i++) {
        std::vector<std::string> v = {"abc", "def"};
        absl::Hash<std::vector<std::string>> hasher;
        hasher(v);
    }
`

for absl::Hash
`for (int i = 0; i < count; i++) {
        std::vector<std::string> v = {"abc", "def"};
        XXH64_hash_t ret = 0;
        for (auto s : v) {
            ret = XXH64(s.c_str(), s.size() + 1, ret);
        }
 }`

And execute those simple code with `time` cmd:
For absl::Hash:
real    0m1.008s
user    0m1.008s
sys     0m0.000s

For xxhash
real    0m1.412s
user    0m1.408s
sys     0m0.004s

It looks like absl::Hash is better on performance.

Reply:
your test is pa good start. There is a nice framework for this. See test/common/http/*_speed_test.cc and refs in the BUILD file for examples.

one slight issue with your test is `for (auto s : v)` looks like it might make a superfluous string copy which isn't fair to xxhash64. try `for (auto& s : v)`. On the other hand, you are only hashing the individual strings, and are not includingthe overhead to build the xxhash64 context for the combined hash, like you had in your earlier commit.

The other issue with your test is that in no case do you need to be making std::string for these, so that adds a common overhead to both sides. use `InlineVector<string_view>`.

One final point on perf tests: it's good when reporting results to indicate how you compiled. `blaze build -c opt` is what we'd like to see for the reported results, though there may be other options you might use if you are trying to do profiling, etc.

But Matt's larger point, I believe, is that he wants hash values to be stable across binary upgrades, which puts absl hash at a functional disadvantage for this use-case. Still I think it's good to understand what the perf implication is, so thank you for getting started with that.

Reply:
> your test is a good start. There is a nice framework for this. See test/common/http/*_speed_test.cc and refs in the BUILD file for examples.

thanks, I will take a look
> one slight issue with your test is `for (auto s : v)` looks like it might make a superfluous string copy which isn't fair to xxhash64. try `for (auto& s : v)`. On the other hand, you are only hashing the individual strings, and are not includingthe overhead to build the xxhash64 context for the combined hash, like you had in your earlier commit.

Thanks again, you are right. Quick test that, I got a pretty close result for xxhash to absl::Hash
`real    0m1.055s
user    0m1.055s
sys     0m0.000s`

> 
> The other issue with your test is that in no case do you need to be making std::string for these, so that adds a common overhead to both sides. use `InlineVector<string_view>`.

Yea, I can test that also.
> 
> One final point on perf tests: it's good when reporting results to indicate how you compiled. `blaze build -c opt` is what we'd like to see for the reported results, though there may be other options you might use if you are trying to do profiling, etc.
> 

Yea, I use  `blaze build -c opt`

> But Matt's larger point, I believe, is that he wants hash values to be stable across binary upgrades, which puts absl hash at a functional disadvantage for this use-case. Still I think it's good to understand what the perf implication is, so thank you for getting started with that.

Totally agree with you, we should understand Matt's point first.

Reply:
Yeah the idea behind using xxhash is to have guaranteed hash stability during upgrades. This has actually caused the project significant issues in the past so IMO it's important that any consistent hash implementations for load balancing are stable across upgrades unless we make an explicit change.

Reply:
Makes sense. Given that, let's just get the best perf we can with xxhash64. You can probably drop both vectors, and just iterate over the headers combining the hash, per your earlier version.